File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/repository/transformer/RxGistsListTransformer.java
package ru.avystavkin.githubgists.repository.transformer;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;
import ru.avystavkin.githubgists.models.local.Gist;
import ru.avystavkin.githubgists.models.server.GistServer;

public class RxGistsListTransformer implements ObservableTransformer<List<GistServer>, List<Gist>> {

    @Override
    public ObservableSource<List<Gist>> apply(Observable<List<GistServer>> observable) {
        return observable
                .flatMap( Observable::fromIterable)
                .map(TransformerHelper::convertGistServer)
                .toList()
                .flatMapObservable(Observable::fromArray);
    }
}

File: GithubGists/app/src/main/java/ru/avystavkin/githubgists/repository/github/transformer/RxGistsListTransformer.kt
package ru.avystavkin.githubgists.repository.github.transformer

import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import ru.avystavkin.githubgists.models.database.Gist
import ru.avystavkin.githubgists.models.server.GistServer

class RxGistsListTransformer : ObservableTransformer<List<GistServer>, List<Gist>> {

    override fun apply(observable: Observable<List<GistServer>>): ObservableSource<List<Gist>> {
        return observable
                .flatMap { x -> Observable.fromIterable(x) }
                .map(TransformerHelper::convertGistServer)
                .toList()
                .flatMapObservable { x -> Observable.fromArray(x) }
    }
}
